### PR TITLE
Tempto runner

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java [hunk 1]
```diff
@@ -81,7 +87,23 @@ public ConventionBasedTest generateProxy(ConventionBasedTest conventionBasedTest
 
     private String generatedClassName(ConventionBasedTest conventionBasedTest)
     {
-        return testPackage + "." + capitalize(replace(conventionBasedTest.testName(), "-", "_"));
+        List<String> testNameParts = Splitter.on('.').splitToList(conventionBasedTest.getTestName());
+        return testPackage + "." + toJavaSymbol(testNameParts.get(testNameParts.size() - 2));
```
Comment (sopel39) [inline tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java]: what is testNameParts.get(testNameParts.size() - 2)?
Extract this as a separate variable?
Is it always 2?

File: tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java [hunk 2]
```diff
@@ -81,7 +87,23 @@ public ConventionBasedTest generateProxy(ConventionBasedTest conventionBasedTest
...
+
+    private String generatedMethodName(ConventionBasedTest conventionBasedTest)
+    {
+        List<String> testNameParts = Splitter.on('.').splitToList(conventionBasedTest.getTestName());
+        return toJavaSymbol(Iterables.getLast(testNameParts));
+    }
+
+    private String toJavaSymbol(String s)
+    {
+        String javaSymbol = s.replaceAll("[^A-Z-a-z0-9_]", "");
```
Comment (sopel39) [inline tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java]: do I understand correctly that '-' is accepted as a java symbol character? In such case I think it is incorrect.

**Comment by kokosing** [comment]
generally it looks ok % comments

File: tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithName.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.teradata.tempto.testmarkers;
+
+import org.testng.ITest;
+
+public interface WithName
+        extends ITest
```
Comment (sopel39) [inline tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithName.java]: where is the getName method? Maybe just test if something implements ITest?

File: tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithTestGroups.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+package com.teradata.tempto.testmarkers;
+
+import org.testng.ITest;
+
+import java.util.Set;
+
+public interface       WithTestGroups
```
Comment (sopel39) [inline tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithTestGroups.java]: format

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        setupTestsConfiguration();
+        TestNG testNG = new TestNG();
+        testNG.setXmlSuites(singletonList(testSuite));
+        setupTestsFiltering(testNG);
+        testNG.run();
+    }
+
+    private void setupTestsConfiguration()
+    {
+        System.setProperty(TestConfigurationFactory.TEST_CONFIGURATION_URI_KEY, options.getConfigFile());
```
Comment (sopel39) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: static import

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        System.setProperty(TestConfigurationFactory.LOCAL_TEST_CONFIGURATION_URI_KEY, options.getConfigFileLocal());
+    }
+
+    private void setupTestsFiltering(TestNG testNG)
+    {
+        if (!options.getTestGroups().isEmpty()) {
+            System.setProperty(TestNameGroupNameMethodSelector.TEST_GROUPS_TO_RUN_PROPERTY, Joiner.on(',').join(options.getTestGroups()));
```
Comment (sopel39) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: static import

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        if (!options.getExcludeGroups().isEmpty()) {
+            System.setProperty(TestNameGroupNameMethodSelector.TEST_GROUPS_TO_EXCLUDE_PROPERTY, Joiner.on(',').join(options.getExcludeGroups()));
+        }
+        if (!options.getTests().isEmpty()) {
+            System.setProperty(TestNameGroupNameMethodSelector.TEST_NAMES_TO_RUN_PROPERTY, Joiner.on(',').join(options.getTests()));
+        }
+        testNG.addMethodSelector("com.teradata.tempto.internal.listeners.TestNameGroupNameMethodSelector", 20);
```
Comment (sopel39) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: 20 is magic constant, extract

Comment (sopel39) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: also, don't use string class name, use TestNameGroupNameMethodSelector Class object and get class name from it.

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        XmlSuite testSuite = new XmlSuite();
+        testSuite.setName("tempto-tests");
+        testSuite.setFileName("tempto-tests");
+        XmlTest test = new XmlTest(testSuite);
+        test.setName("all");
+        XmlPackage testPackage = new XmlPackage(options.getTestsPackage());
+        List<XmlPackage> testPackages = new ArrayList<>();
+        testPackages.add(testPackage);
+        test.setPackages(testPackages);
+        XmlClass conventionBasedTestsClass = new XmlClass("com.teradata.tempto.internal.convention.ConventionBasedTestFactory");
```
Comment (sopel39) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: use Class object to get class name

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunnerCommandLineParser.java [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    private static final String REPORT_DIR_OPTION = "report-dir";
+    private static final String REPORT_DIR_DEFAULT = "./test-reports";
+
+    private static final String GROUPS_OPTION = "groups";
+    private static final String EXCLUDED_GROUPS_OPTION ="excluded-groups";
+    private static final String TESTS_OPTION = "tests";
+
+
+    private static final String HELP_OPTION = "help";
+    public static final Comparator<Option> ALL_EQUAL_OPTION_COMPARATOR = (a, b) -> 0;
```
Comment (sopel39) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunnerCommandLineParser.java]: this doesn't do what is says, rename it to always equal comparator or smth

**Comment by sopel39** [comment]
seems ok & comments
& update docs & notify @brian-rickman about this changes (e.g. python runner removed)

File: tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithName.java (hunk 3 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithName.java]: WithName is more readable.

File: tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithTestGroups.java (hunk 4 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithTestGroups.java]: done

File: tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithName.java (hunk 3 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/testmarkers/WithName.java]: added explicit getTestName method declaration.

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java (hunk 5 above)
Comment (losipiuk) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: done

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java (hunk 6 above)
Comment (losipiuk) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: done

File: tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java (hunk 7 above)
Comment (losipiuk) [inline tempto-runner/src/main/java/com/teradata/tempto/runner/TemptoRunner.java]: done and done

File: tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java (hunk 2 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java]: you are right - fixed

File: tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java (hunk 1 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/convention/ConventionBasedTestProxyGenerator.java]: added test

**Comment by losipiuk** [comment]
I think I addressed everything. Please take a look at AC commits and let me know if I can push it. 
I will squash beforehead :)

**Pull request closed without merging**